Source/WebDriver/WebSocketServer.h

```diff
@@ -0,0 +1,146 @@
+};
+
+
+class WebSocketMessageHandler {
+public:
+
+#if USE(SOUP)
+    typedef GRefPtr<SoupWebsocketConnection> Connection;
+#endif
+
+    struct Message {
+        // Optional connection, as the message might be a reply to a command not yet associated to a connection
+        std::optional<Connection> connection;
+        const char* data { nullptr };
+        size_t dataLength { 0 };
+
+        static Message fail(CommandResult::ErrorCode, std::optional<Connection>, std::optional<String> errorMessage = std::nullopt, std::optional<String> commandId = std::nullopt);
+    };
+
+    virtual void handleHandshake(HTTPRequestHandler::Request&&, Function<void(std::optional<HTTPRequestHandler::Response>&& response)>&& replyHandler) = 0;
+    virtual void handleMessage(Message&&, Function<void(Message&&)>&& replyHandler) = 0;
+    virtual void clientDisconnected(Connection) = 0;
+private:
+};
+
+class WebSocketServer : public RefCounted<WebSocketServer>, public CanMakeWeakPtr<WebSocketServer> {
+public:
+    explicit WebSocketServer(WebSocketMessageHandler&, WebDriverService&);
+    virtual ~WebSocketServer() = default;
+
+    std::optional<String> listen(const std::optional<String>& host, unsigned port);
+    bool isListening();
+    void disconnect();
+
+    WebSocketMessageHandler& messageHandler() { return m_messageHandler; }
+
+    const RefPtr<WebSocketListener> listener() const { return m_listener; }
+    void addToConnectionsNotAssociatedToASession(WebSocketMessageHandler::Connection);
+    void addToConnectionsAssociatedToASession(std::pair<WebSocketMessageHandler::Connection, String>);
+    bool isConnectionNotAssociatedToSession(WebSocketMessageHandler::Connection&);
+    void removeConnectionNotAssociatedWithSession(WebSocketMessageHandler::Connection&);
+    std::optional<RefPtr<Session>> getSessionForConnection(WebSocketMessageHandler::Connection&);
+    std::optional<WebSocketMessageHandler::Connection> getConnectionForSession(const String& sessionId);
+    void removeConnectionAssociatedToSession(WebSocketMessageHandler::Connection&);
+    RefPtr<WebSocketListener> startListeningForAWebSocketConnectionGivenSession(String sessionId);
+    String constructWebSocketResourceNameGivenSession(String sessionId);
+    String constructWebSocketURLForListenerAndSession(const RefPtr<WebSocketListener>, String sessionId);
+    String getSessionIDForWebSocketResource(const String& resource);
+    void sendMessageFromSession(const String& session, const String& message);
+
+    // Non-spec method
+    void removeResourceForSession(const String& sessionId);
+    void disconnectSession(const String& sessionId);
+
+private:
+
+    WebSocketMessageHandler& m_messageHandler;
+    WebDriverService& m_service;
+    String m_listenerURL;
+    RefPtr<WebSocketListener> m_listener;
+    // FIXME GRefPtr (our Connection type in Soup is failing to be used as a std::map key, so we use a simple pair for now)
```

Comment:
Nit: add bugzilla link

Reply:
Replaced with comment about the decision to go with `std::vector` instead of `std::map` for now.